src/Microsoft.VisualStudio.ProjectSystemTools.BackEnd/BackEnd/RoslynLogger.cs

```diff
@@ -0,0 +1,262 @@
+                    if (type == null)
+                    {
+                        return false;
+                    }
+
+                    MethodInfo? setLoggerMethodInfo = type.GetMethod("SetLogger", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public);
+                    if (setLoggerMethodInfo != null)
+                    {
+                        _setLoggerCall = (Action<TraceSource>)setLoggerMethodInfo.CreateDelegate(typeof(Action<TraceSource>));
+                    }
+
+                    MethodInfo? removeLoggerMethodInfo = type.GetMethod("RemoveLogger", BindingFlags.Static | BindingFlags.NonPublic | BindingFlags.Public);
+                    if (removeLoggerMethodInfo != null)
+                    {
+                        _removeLoggerCall = (Action<TraceSource>)removeLoggerMethodInfo.CreateDelegate(typeof(Action<TraceSource>));
+                    }
+                }
+                catch
+                {
+                    return false;
+                }
+            }
+
+            return _setLoggerCall != null;
+        }
+
+        private static Assembly GetAssembly(string assemblyName) =>
+            (from assembly in AppDomain.CurrentDomain.GetAssemblies()
+                where assembly.FullName.IndexOf(assemblyName, StringComparison.OrdinalIgnoreCase) >= 0
+                let current = assembly.GetName()
+                where string.Equals(current.Name, assemblyName, StringComparison.OrdinalIgnoreCase)
+                select assembly).FirstOrDefault();
+
+        private class RoslynTraceListener : TraceListener
+        {
+            private const int LogEventId = 0;
+            private const int StartEventId = 1;
+            private const int EndEventId = 2;
+
+            private readonly IImmutableSet<string> _set;
+            private readonly StreamWriter? _writer;
+            private readonly object _writerLock = new object();
+
+            public RoslynTraceListener(IImmutableSet<string> roslynEvents)
+            {
+                LogPath = Path.Combine(Path.GetTempPath(), $"RoslynLog-{Guid.NewGuid()}.txt");
+
+                _set = roslynEvents;
+
+                try
+                {
+                    // if there is an issue, this will crash VS. we can swallow, but then, user won't know
+                    // why logging didn't work.
+                    _writer = new StreamWriter(LogPath);
+                }
+                catch
+                {
+                    // Prevent server VS from crashing
+                    System.Diagnostics.Debug.Assert(false);
+                }
+            }
+
+            public readonly string LogPath;
+            public override bool IsThreadSafe => true;
+
+            public override void TraceData(TraceEventCache eventCache, string source, TraceEventType eventType, int id, params object[] data)
+            {
+                string? functionId = (string)data[0];
```

Comment:
Can this be null? If so, `data` should have type `object?[]`. If not, remove the `?` here.